src/devices/machine/spi_sdcard.cpp

```diff
@@ -121,69 +129,85 @@ void spi_sdcard_device::spi_clock_w(int state)
 			if (m_cur_bit == 8)
 			{
 				LOGMASKED(LOG_SPI, "SDCARD: got %02x\n", m_in_latch);
+				for (u8 i = 0; i < 5; i++)
+				{
+					m_cmd[i] = m_cmd[i + 1];
+				}
+				m_cmd[5] = m_in_latch;
 
 				switch (m_state)
 				{
-					case SD_STATE_IDLE:
-						for (int i = 0; i < 5; i++)
+				case SD_STATE_IDLE:
+					do_command(m_cmd);
+					break;
+
+				case SD_STATE_WRITE_WAITFE:
+					if (m_in_latch == 0xfe)
+					{
+						m_state = SD_STATE_WRITE_DATA;
+						m_out_latch = 0xff;
+						m_write_ptr = 0;
+						change_state(SD_STATE_RCV);
+					}
+					break;
+
+				case SD_STATE_WRITE_DATA:
+					m_data[m_write_ptr++] = m_in_latch;
+					if (m_write_ptr == (m_blksize + 2))
+					{
+						u32 blk = ((u32) m_cmd[1] << 24) | ((u32) m_cmd[2] << 16) | ((u32) m_cmd[3] << 8) | (u32) m_cmd[4];
```

Comment:
Please use function call style casts – they’re less likely to need additional parentheses and less error-prone.